geode-core/src/main/java/org/apache/geode/internal/cache/CacheDistributionAdvisor.java

```diff
@@ -145,14 +153,22 @@ public String toString() {
   private Set<InternalDistributedMember> adviseAllEventsOrCached(final boolean excludeInRecovery)
       throws IllegalStateException {
     getAdvisee().getCancelCriterion().checkCancelInProgress(null);
-    return adviseFilter(profile -> {
-      assert profile instanceof CacheProfile;
-      CacheProfile cp = (CacheProfile) profile;
-      if (excludeInRecovery && cp.inRecovery) {
-        return false;
-      }
-      return cp.cachedOrAllEventsWithListener();
-    });
+
+    if (adviseAllEventsVersion != profilesVersion || adviseAllEventsVersion == 0
+        || excludeInRecovery != prevExcludeInRecovery) {
+      adviseSetforAllEvents = Collections.unmodifiableSet(adviseFilter(profile -> {
+        assert profile instanceof CacheProfile;
+        CacheProfile cp = (CacheProfile) profile;
+        if (excludeInRecovery && cp.inRecovery) {
+          return false;
+        }
+        return cp.cachedOrAllEventsWithListener();
+      }));
+      adviseAllEventsVersion = profilesVersion;
```

Comment:
Between the above check and this update the `profileVersion` can be concurrently updated by calls to `basicAdd/Remove` methods.